samza-api/src/main/java/org/apache/samza/task/TaskContext.java

```diff
@@ -54,4 +54,14 @@
    *
    */
   void setStartingOffset(SystemStreamPartition ssp, String offset);
+
+  /**
+   * Method to allow user to return customized context
+   *
+   * @param <T>  the type of user-defined task context
+   * @return  user-defined task context object
+   */
+  default <T> T getUserDefinedContext() {
```

Comment:
No test yet. Thanks for testing! I agree that the type T here does not provide much value. Will remove it.